### ISSUE TITLE
Register guides in AutoGuideList as submodules

### ISSUE THREAD

**Issue by neerajprad** [pyro-ppl/pyro#2131]
After #2126, autoguides subclass from `nn.Module` and support all their functionality. The design for `AutoGuideList` can also be made more module friendly. Currently, in `_setup_prototype` we iterate over all the guides and add set their parameters on to the main class so that these are detectable by `pyro.module`. 

Instead, we could add the logic from `AutoGuideList.add` to `__setattr__` in `ConstrainedModule`. Then, we will be able to directly set the subcomponent guides as `autoguidelist.next_guide = ...`. This will register `next_guide` as a submodule and make all its parameters available via `named_parameters()` without us having to do any additional setup in `_setup_prototype`. 

More importantly, currently any guide that itself contains a submodule (we don't have any currently) would throw an error with `AutoGuideList`. The result is that the submodule parameters will be named as `submodule_name.param_name` and as such `setattr(self, "submodule_name.param_name", value)` will fail in `_setup_prototype`. Registering these guides as submodules is much safer since there won't be name clashes and we can get rid of the hacky `setattr` logic in `_setup_prototype`.

**Comment by fritzo** [comment]
Should we aim to make this change by 1.0 release?
Should we fix the "prefix" issue by 1.0 release?

**Comment by neerajprad** [comment]
> Should we fix the "prefix" issue by 1.0 release?

I am hoping that this should be a relatively small fix, and we can do it before 1.0. I can take a look. It will be nice to have this by 1.0.

**Comment by fritzo** [comment]
Closed by #2137
